packages/core/src/utils/filesearch/fileSearch.ts

```diff
@@ -0,0 +1,194 @@
+  constructor(message = 'Search aborted') {
+    super(message);
+    this.name = 'AbortError';
+  }
+}
+
+export async function filter(
+  allPaths: string[],
+  pattern: string,
+  signal: AbortSignal | undefined,
+): Promise<string[]> {
+  const isGlob = pattern.includes('*');
+  let finalPattern = pattern;
+
+  if (isGlob && !pattern.includes('**')) {
+    finalPattern = `**/${pattern}`;
+  }
+
+  const patternFilter = isGlob
+    ? picomatch(finalPattern, { posix: true, dot: true })
+    : (filePath: string) => {
+        try {
+          return new RegExp(pattern, 'i').test(filePath);
+        } catch (error: unknown) {
+          if (error instanceof Error) {
+            console.error(`Error testing pattern: ${error.message}`);
+          }
+
+          return false;
+        }
+      };
+
+  const results: string[] = [];
+  for (const [i, p] of allPaths.entries()) {
+    if (i % 1000 === 0) {
+      await new Promise((resolve) => {
+        setImmediate(resolve);
+      });
+      if (signal?.aborted) {
+        throw new AbortError();
+      }
+    }
+
+    const normalizedPath = p.replace(/\\/g, '/');
+    const matches = patternFilter(normalizedPath);
+    if (matches) {
+      results.push(p);
+    }
+  }
+
+  results.sort((a, b) => {
+    const aIsDir = a.endsWith('/') || a.endsWith('\\');
+    const bIsDir = b.endsWith('/') || b.endsWith('\\');
+
+    if (aIsDir && !bIsDir) return -1;
+    if (!aIsDir && bIsDir) return 1;
+    return a.localeCompare(b);
+  });
+
+  return results;
+}
+
+export class FileSearch {
+  private readonly absoluteDir: string;
+  private readonly ignore: Ignore = new Ignore();
+  private memoryCache: MemoryCache | undefined;
+  private allFiles: string[] = [];
+
+  constructor(private readonly options: FileSearchOptions) {
+    this.absoluteDir = path.resolve(options.projectRoot);
+  }
+
+  async initialize(): Promise<void> {
+    this.loadIgnoreRules();
+    await this.crawlFiles();
+    this.buildMemoryCache();
+  }
+
+  async search(pattern: string, signal?: AbortSignal): Promise<string[]> {
+    if (!this.memoryCache) {
+      throw new Error('Engine not initialized.');
+    }
+
+    if (pattern === '') {
+      return this.memoryCache!.get('*/', signal);
+    }
+
+    return this.memoryCache!.get(pattern, signal);
+  }
+
+  private loadIgnoreRules(): void {
+    if (this.options.useGitignore) {
+      const gitignorePath = path.join(this.absoluteDir, '.gitignore');
+      if (fs.existsSync(gitignorePath)) {
+        this.ignore.add(fs.readFileSync(gitignorePath, 'utf8'));
+      }
+    }
+
+    if (this.options.useGeminiignore) {
+      const geminiignorePath = path.join(this.absoluteDir, '.geminiignore');
+      if (fs.existsSync(geminiignorePath)) {
+        this.ignore.add(fs.readFileSync(geminiignorePath, 'utf8'));
+      }
+    }
+
+    if (this.options.ignoreDirs.length > 0) {
+      this.ignore.add(
+        this.options.ignoreDirs.map((dir) => {
+          if (dir.endsWith('/')) {
```

Comment:
make sure we use the appropriate path separators for the current operating system.